### PR TITLE
Enable kapt support

### PR DESCRIPTION
**Pull request by thalescm**
Note that this PR depends on the [actual buck kapt support](https://github.com/facebook/buck/pull/1584) to be merged

### CODE REVIEW

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidTarget.groovy [hunk 1]
```diff
@@ -141,14 +142,20 @@ abstract class AndroidTarget extends JavaLibTarget {
...
     Scope getTestApt() {
-        return Scope.from(project, unitTestVariant
-                ? unitTestVariant.annotationProcessorConfiguration : null
-        )
+        if (!isKotlin)
+            return Scope.from(project, unitTestVariant
+                    ? unitTestVariant.annotationProcessorConfiguration : null
+            )
+        else
+            return Scope.from(project, KotlinLibTarget.ANNOTATION_PROCESSOR_CONFIGURATION_NAME)
```
Comment (runningcode) [inline buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidTarget.groovy]: Could you add a test to make sure that UnitTests in modules with KAPT work? (if one doesn't already exist)

File: dependencies.gradle [hunk 2]
```diff
@@ -44,6 +44,7 @@ def apt = [
 def external = [
         butterKnife    : "com.jakewharton:butterknife:${versions.butterKnife}",
         dagger         : "com.google.dagger:dagger:${versions.dagger}",
+        daggerAndroid  : "com.google.dagger:dagger-android-support:${versions.dagger}",
```
Comment (runningcode) [inline dependencies.gradle]: is the dagger android compiler required for this to work properly?

File: libraries/kotlinlibrary/src/main/java/demo/di/Coffee.kt [hunk 3]
```diff
@@ -0,0 +1,81 @@
+package demo.di
+
+import dagger.*
```
Comment (runningcode) [inline libraries/kotlinlibrary/src/main/java/demo/di/Coffee.kt]: star import!

File: buildSrc/build.gradle [hunk 4]
```diff
@@ -73,7 +73,7 @@ task sourcesJar(type: Jar) {
     classifier = "sources"
 }
 
-def publishVersion = "0.34.1"
+def publishVersion = "0.35.0"
```
Comment (kageiit) [inline buildSrc/build.gradle]: Please undo this change

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidTarget.groovy [hunk 5]
```diff
@@ -141,14 +142,20 @@ abstract class AndroidTarget extends JavaLibTarget {
 
     @Override
     Scope getApt() {
-        return Scope.from(project, baseVariant.annotationProcessorConfiguration)
+        if (!isKotlin)
+            return Scope.from(project, baseVariant.annotationProcessorConfiguration)
```
Comment (kageiit) [inline buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidTarget.groovy]: can you inline the if condition please

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidTarget.groovy [hunk 6]
```diff
@@ -141,14 +142,20 @@ abstract class AndroidTarget extends JavaLibTarget {
...
+            return Scope.from(project, KotlinLibTarget.ANNOTATION_PROCESSOR_CONFIGURATION_NAME)
     }
 
     @Override
     Scope getTestApt() {
-        return Scope.from(project, unitTestVariant
-                ? unitTestVariant.annotationProcessorConfiguration : null
-        )
+        if (!isKotlin)
+            return Scope.from(project, unitTestVariant
```
Comment (kageiit) [inline buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidTarget.groovy]: can you inline the if condition please

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/kotlin/KotlinLibTarget.groovy [hunk 7]
```diff
@@ -21,4 +24,14 @@ class KotlinLibTarget extends JavaLibTarget {
...
     }
+
+    @Override
+    Scope getApt() {
+        return Scope.from(project, ANNOTATION_PROCESSOR_CONFIGURATION_NAME)
+    }
+
+    @Override
+    Scope getTestApt() {
+        return Scope.from(project, ANNOTATION_PROCESSOR_CONFIGURATION_NAME)
```
Comment (kageiit) [inline buildSrc/src/main/groovy/com/uber/okbuck/core/model/kotlin/KotlinLibTarget.groovy]: This should be testKapt?

Ideally you want to use the same constants the kotlin gradle plugin uses. Can you please check if they have the constants available publicly and switch to using those instead?

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidTarget.groovy [hunk 8]
```diff
@@ -141,14 +142,20 @@ abstract class AndroidTarget extends JavaLibTarget {
 
     @Override
     Scope getApt() {
-        return Scope.from(project, baseVariant.annotationProcessorConfiguration)
+        if (!isKotlin)
+            return Scope.from(project, baseVariant.annotationProcessorConfiguration)
+        else
+            return Scope.from(project, KotlinLibTarget.ANNOTATION_PROCESSOR_CONFIGURATION_NAME)
```
Comment (kageiit) [inline buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidTarget.groovy]: This seems like a bad code smell. Android target depending on a constant from kotlin target.

Lets move constant outside or use the ones defined by the kotlin gradle plugin api directly

File: libraries/kotlinlibrary/build.gradle [hunk 9]
```diff
@@ -1,7 +1,21 @@
...
+  implementation deps.external.dagger
+  kapt deps.apt.daggerCompiler
+  testImplementation deps.test.junit
+  testImplementation deps.test.kotlinTest
+}
+
+// At the moment for non-Android projects you need to explicitly
+// mark the code generated by kapt as 'generated source code'
+// for correct highlighting and resolve in IDE.
+idea {
```
Comment (kageiit) [inline libraries/kotlinlibrary/build.gradle]: We dont import this with gradle, so we dont need this. Please remove

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/kotlin/KotlinLibTarget.groovy (hunk 7 above)
Comment (thalescm) [reply buildSrc/src/main/groovy/com/uber/okbuck/core/model/kotlin/KotlinLibTarget.groovy]: They do not, searched inside kotlin-gradle-plugin project. Only place I've found kapt written was in [here](https://github.com/JetBrains/kotlin/blob/6838873a93f2fc33ccdbaea5788e08fc154ac2d1/gradle/project-schema.json) 😢

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidTarget.groovy (hunk 8 above)
Comment (thalescm) [reply buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidTarget.groovy]: This should actually probably use the `baseVariant`, but I honestly don't know how `annotationProcessorConfiguration` is built

File: dependencies.gradle (hunk 2 above)
Comment (thalescm) [reply dependencies.gradle]: I can remove this dependency and make a custom module, similar to `Coffee.kt` added on `kotlinlibrary` project.
Just added this for a more "real looking" project

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/kotlin/KotlinLibTarget.groovy (hunk 7 above)
Comment (kageiit) [reply buildSrc/src/main/groovy/com/uber/okbuck/core/model/kotlin/KotlinLibTarget.groovy]: They have it defined in https://github.com/JetBrains/kotlin/blob/1f81c0cdfe9d9438667094d01f2bd69f0d6fa780/libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/internal/Kapt3KotlinGradleSubplugin.kt#L79

but its internal.

You should be careful because there are variant specific kapt configurations as well for android projects which this pr does not capture correctly

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidInstrumentationTarget.groovy [hunk 10]
```diff
@@ -24,6 +24,9 @@ class AndroidInstrumentationTarget extends AndroidAppTarget {
 
     @Override
     Scope getApt() {
+        if (isKapt) {
+            return Scope.from(project, "kaptAndroidTest")
```
Comment (kageiit) [inline buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidInstrumentationTarget.groovy]: This does not handle variant specific configurations correctly

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidTarget.groovy [hunk 11]
```diff
@@ -141,14 +145,18 @@ abstract class AndroidTarget extends JavaLibTarget {
...
         return Scope.from(project, baseVariant.annotationProcessorConfiguration)
     }
 
     @Override
     Scope getTestApt() {
-        return Scope.from(project, unitTestVariant
-                ? unitTestVariant.annotationProcessorConfiguration : null
-        )
+        if (isKapt)  {
+            return Scope.from(project, "kapt${baseVariant.getName().capitalize()}")
```
Comment (kageiit) [inline buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidTarget.groovy]: use `baseVariant.name` like before instead of `getName()`

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidTarget.groovy [hunk 12]
```diff
@@ -141,14 +145,18 @@ abstract class AndroidTarget extends JavaLibTarget {
 
     @Override
     Scope getApt() {
+        if (isKapt) {
+            return Scope.from(project, "kapt${baseVariant.name.capitalize()}")
+        }
         return Scope.from(project, baseVariant.annotationProcessorConfiguration)
```
Comment (kageiit) [inline buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidTarget.groovy]: Can you pleas inline like the following

```
return Scope.from(project, isKapt ? "kapt${baseVariant.name.capitalize()}" : baseVariant.annotationProcessorConfiguration)
```

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidTarget.groovy [hunk 13]
```diff
@@ -141,14 +145,18 @@ abstract class AndroidTarget extends JavaLibTarget {
...
     }
 
     @Override
     Scope getTestApt() {
-        return Scope.from(project, unitTestVariant
-                ? unitTestVariant.annotationProcessorConfiguration : null
-        )
+        if (isKapt)  {
```
Comment (kageiit) [inline buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidTarget.groovy]: inline

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidInstrumentationTarget.groovy (hunk 10 above)
Comment (thalescm) [reply buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidInstrumentationTarget.groovy]: https://youtrack.jetbrains.com/issue/KT-23411

**Comment by thalescm** [comment]
@kageiit if you could review it again, please :)

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidTarget.groovy [hunk 14]
```diff
@@ -17,13 +17,15 @@ import com.uber.okbuck.core.model.base.RuleType
...
 import com.uber.okbuck.core.model.jvm.TestOptions
+import com.uber.okbuck.core.model.kotlin.KotlinLibTarget
 import com.uber.okbuck.core.util.FileUtil
 import groovy.util.slurpersupport.GPathResult
 import groovy.xml.StreamingMarkupBuilder
 import org.gradle.api.JavaVersion
 import org.gradle.api.Project
 import org.gradle.api.tasks.compile.JavaCompile
 import org.gradle.api.tasks.testing.Test
+import org.jetbrains.kotlin.gradle.internal.Kapt3GradleSubplugin
```
Comment (kageiit) [inline buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidTarget.groovy]: lets not use internal classes

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidTarget.groovy [hunk 15]
```diff
@@ -91,6 +94,7 @@ abstract class AndroidTarget extends JavaLibTarget {
 
         // Check if kotlin
         isKotlin = project.plugins.hasPlugin(KotlinAndroidPluginWrapper.class)
+        isKapt = project.plugins.hasPlugin(Kapt3GradleSubplugin.class)
```
Comment (kageiit) [inline buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidTarget.groovy]: lets not use internal classes

just check if the `kapt` string is applied

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidTarget.groovy [hunk 16]
```diff
@@ -141,14 +145,15 @@ abstract class AndroidTarget extends JavaLibTarget {
 
     @Override
     Scope getApt() {
-        return Scope.from(project, baseVariant.annotationProcessorConfiguration)
+        return Scope.from(project,
+                isKapt ?"kapt${baseVariant.name.capitalize()}" : baseVariant.annotationProcessorConfiguration)
```
Comment (kageiit) [inline buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidTarget.groovy]: space after ?

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/kotlin/KotlinLibTarget.groovy [hunk 17]
```diff
@@ -1,7 +1,9 @@
 package com.uber.okbuck.core.model.kotlin
 
+import com.uber.okbuck.core.model.base.Scope
 import com.uber.okbuck.core.model.java.JavaLibTarget
 import org.gradle.api.Project
+import org.jetbrains.kotlin.gradle.internal.Kapt3GradleSubplugin
```
Comment (kageiit) [inline buildSrc/src/main/groovy/com/uber/okbuck/core/model/kotlin/KotlinLibTarget.groovy]: unused import

File: kotlin-app/.gitignore [hunk 18]
```diff
@@ -0,0 +1 @@
+/build
```
Comment (kageiit) [inline kotlin-app/.gitignore]: remove this file

File: kotlin-app/build.gradle [hunk 19]
```diff
@@ -0,0 +1,38 @@
+apply plugin: "com.android.application"
+apply plugin: 'kotlin-android'
```
Comment (kageiit) [inline kotlin-app/build.gradle]: use `"` consistently in the gradle file everywhere instead of `'` for strings

File: kotlin-app/proguard-rules.pro [hunk 20]
```diff
@@ -0,0 +1,21 @@
+# Add project specific ProGuard rules here.
```
Comment (kageiit) [inline kotlin-app/proguard-rules.pro]: delete this file

**Comment by kageiit** [comment]
Lets change buckVersion in OkBuckExtension to latest i.e `8ce226cd2da6ac107d7db9535d6db4a1ac16211a` so we have the kapt bits available in buck
